src/nomad_catalysis/schema_packages/schema.py

```diff
@@ -0,0 +1,424 @@
+    for attr in attr_path.split('.'):
+        obj = getattr(obj, attr, None)
+        if obj is None:
+            return None
+    return obj
+
+
+def set_nested_attr(obj, attr_path, value):
+    for attr in attr_path.split('.'):
+        obj = getattr(obj, attr, None)
+        if obj is None:
+            return
+    setattr(obj, attr[-1], value)
+
+
+def populate_catalyst_sample_info_to_results(archive, self, logger, reference=False):  # noqa: PLR0912
+    """
+    Copies the catalyst sample information (if reference provided from a reference)
+    into the results archive of the measurement.
+    """
+    if reference:
+        sample_obj = self.samples[0].reference
+    else:
+        sample_obj = self
+
+    add_catalyst(archive)
+    quantities_results_mapping = {
+        'name': 'catalyst_name',
+        'catalyst_type': 'catalyst_type',
+        'preparation_details.preparation_method': 'preparation_method',
+        'surface.surface_area': 'surface_area',
+        'surface.method_surface_area_determination': 'characterization_methods',
+    }
+
+    # Loop through the mapping and assign the values
+    for ref_attr, catalyst_attr in quantities_results_mapping.items():
+        value = get_nested_attr(sample_obj, ref_attr)
+        if value is not None:
+            try:
+                setattr(
+                    archive.results.properties.catalytic.catalyst, catalyst_attr, value
+                )
+            except ValueError:
+                setattr(
+                    archive.results.properties.catalytic.catalyst,
+                    catalyst_attr,
+                    [value],
+                )
+    if reference:
+        if self.samples[0].reference.name is not None:
+            if not archive.results.material:
+                archive.results.material = Material()
+            archive.results.material.material_name = self.samples[0].reference.name
+
+        if self.samples[0].reference.elemental_composition is not None:
+            if not archive.results.material:
+                archive.results.material = Material()
+            try:
+                archive.results.material.elemental_composition = self.samples[
+                    0
+                ].reference.elemental_composition
+
+            except Exception as e:
+                logger.warn('Could not analyse elemental compostion.', exc_info=e)
+
+            for i in self.samples[0].reference.elemental_composition:
+                if i.element not in chemical_symbols:
+                    logger.warn(
+                        f"'{i.element}' is not a valid element symbol and this "
+                        'elemental_composition section will be ignored.'
+                    )
+                elif i.element not in archive.results.material.elements:
+                    archive.results.material.elements += [i.element]
+
+
+def add_referencing_methods_to_sample_result(self, archive, logger, number):
+    if self.lab_id is not None:
```

Comment:
This is a bit of an anti-pattern in python that can cause a lot of nesting and hurt the readability. It would be better to check if it is None and then return. Something like:
```py
if self.lab_id is None:
    logger.warn(f'Found no entries with reference: "{catalyst_sample}".')
    return
```
Then you can remove the else at the bottom and reduce the indentation.